### PR TITLE
Infra/picker types arrangement

### PR DESCRIPTION
**Pull request by adids1221**
## Description
Picker types arrangement, new `useNewPickerProps` to convert old props into new props.
Picker types divide into type of the component that uses these props.
Also some props are renamed according to the component that uses them.

## Changelog
Picker types arrangement - new props, props deprecation.

## Additional info
None

### CODE REVIEW

File: src/components/picker/types.tsx [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import React, {PropsWithChildren, ReactNode} from 'react';
 import {FlatListProps, StyleProp, ViewStyle, TextStyle} from 'react-native';
 import {ExpandableOverlayProps, ExpandableOverlayMethods} from '../../incubator/expandableOverlay';
-import {ModalTopBarProps} from '../modal/TopBar';
+import {ModalTopBarProps} from '../../components/modal';
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:4]: Why this change?

File: src/components/picker/types.tsx [hunk 2]
```diff
@@ -147,49 +130,112 @@ export type PickerBaseProps = Omit<NewTextFieldProps, 'value' | 'onChange'> & {
...
+  enableModalBlur?: boolean;
+};
+
+type PickerListProps = {
   /**
-   * Pass props to the picker modal
+   * Render a custom header for Picker's Overlay
    */
-  pickerModalProps?: object;
+  renderCustomOverlayHeader?: (callbacks: {onDone?: () => void; onCancel?: () => void}) => React.ReactElement;
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:154]: Why no simply 'renderHeader'?

File: src/components/picker/types.tsx [hunk 3]
```diff
@@ -147,49 +130,112 @@ export type PickerBaseProps = Omit<NewTextFieldProps, 'value' | 'onChange'> & {
...
+   * Custom picker props (when using renderPicker, will apply on the button wrapper)
    */
-  renderCustomDialogHeader?: (callbacks: {onDone?: () => void, onCancel?: ()=> void}) => React.ReactElement;
+  customPickerProps?: ExpandableOverlayProps;
   /**
-   * Use wheel picker instead of a list picker
+   * Render custom picker modal (e.g ({visible, children, toggleModal}) => {...})
    */
-  useWheelPicker?: boolean;
+  renderPickerOverlay?: (modalProps: RenderCustomModalProps) => React.ReactElement;
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:143]: Why not simply 'renderOverlay' like 'renderInput'?

File: src/components/picker/types.tsx [hunk 4]
```diff
@@ -147,49 +130,112 @@ export type PickerBaseProps = Omit<NewTextFieldProps, 'value' | 'onChange'> & {
...
+     */
+    onPress?: () => void;
+    /**
+     * A function that returns the label to show for the selected Picker value
+     */
+    getLabel?: (value: PickerValue) => string;
+    /**
+     * The picker modal top bar props
+     */
+    topBarProps?: ModalTopBarProps;
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:220]: Why it's not a part of the PickerListProps?

File: src/components/picker/types.tsx [hunk 5]
```diff
@@ -147,49 +130,112 @@ export type PickerBaseProps = Omit<NewTextFieldProps, 'value' | 'onChange'> & {
...
-   */
-  testID?: string;
-  /**
-   * @deprecated
-   */
-  children?: ReactNode | undefined;
 };
 
+export type PickerBaseProps = Omit<NewTextFieldProps, 'value' | 'onChange'> &
+  PickerSearchProps &
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:166]: PickerSearchProps are part of the PickerListProps

File: src/components/picker/types.tsx [hunk 6]
```diff
@@ -48,85 +48,68 @@ export interface PickerSearchStyle {
...
+   * @deprecated
+   * A function that extract the label out of the value prop in case value has a custom structure (e.g. {myValue, myLabel})
    */
-  onChange?: (value: PickerValue) => void;
+  getItemLabel?: (value: PickerValue) => string;
   /**
-   * SINGLE mode or MULTI mode
+   * @deprecated
+   * Callback for modal onShow event
+   * Instead pass onShow via customPickerProps.modalProp | pickerModalProps
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:70]: Don't mention `pickerModalProps` if it's being deprecated

File: src/components/picker/types.tsx [hunk 7]
```diff
@@ -48,85 +48,68 @@ export interface PickerSearchStyle {
...
-   * Render custom picker item
-   */
-  renderItem?: (
-    value: PickerValue,
-    itemProps: PickerItemProps & {isSelected: boolean; isItemDisabled: boolean},
-    label?: string
-  ) => React.ReactElement;
-  /**
+   * @deprecated
    * Render custom picker modal (e.g ({visible, children, toggleModal}) => {...})
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:94]:
```suggestion
   * Render custom picker overlay (e.g ({visible, children, toggleModal}) => {...})
```

File: src/components/picker/types.tsx [hunk 8]
```diff
@@ -48,85 +48,68 @@ export interface PickerSearchStyle {
...
-   * Adds blur effect to picker modal (iOS only)
+   * @deprecated
+   * Render a custom header for Picker's dialog
+   * instead use renderCustomOverlayHeader
    */
-  enableModalBlur?: boolean;
+  renderCustomDialogHeader?: (callbacks: {onDone?: () => void; onCancel?: () => void}) => React.ReactElement;
   /**
+   * @deprecated
    * Render custom picker - input will be value (see above)
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:86]:
```suggestion
   * Render custom picker input (the value will be passed)
```

File: src/components/picker/types.tsx [hunk 9]
```diff
@@ -48,85 +48,68 @@ export interface PickerSearchStyle {
...
-   * A function that returns the label to show for the selected Picker value
-   */
-  getLabel?: (value: PickerValue) => string;
-  /**
-   * The picker modal top bar props
+   * Custom container style
+   * instead use inputContainerStyle
    */
-  topBarProps?: ModalTopBarProps;
+  containerStyle?: StyleProp<ViewStyle>;
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:109]: I don't think 'containerStyle' should be renamed. ContainerStyle is a proper name for style passed to the container, which in this case is the input's container.

File: src/components/picker/types.tsx [hunk 10]
```diff
@@ -147,49 +130,112 @@ export type PickerBaseProps = Omit<NewTextFieldProps, 'value' | 'onChange'> & {
    * Render custom search input (only when passing showSearch)
    */
   renderCustomSearch?: (props: PickerItemsListProps) => React.ReactElement;
+};
+
+type PickerExpandableOverlayProps = {
   /**
-   * Render a custom header for Picker's dialog
+   * Custom picker props (when using renderPicker, will apply on the button wrapper)
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:137]:
```suggestion
   * Custom picker props
```

File: src/components/picker/types.tsx [hunk 11]
```diff
@@ -147,49 +130,112 @@ export type PickerBaseProps = Omit<NewTextFieldProps, 'value' | 'onChange'> & {
...
    */
-  renderCustomDialogHeader?: (callbacks: {onDone?: () => void, onCancel?: ()=> void}) => React.ReactElement;
+  customPickerProps?: ExpandableOverlayProps;
   /**
-   * Use wheel picker instead of a list picker
+   * Render custom picker modal (e.g ({visible, children, toggleModal}) => {...})
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:141]:
```suggestion
   * Render custom picker overlay (e.g ({visible, children, toggleModal}) => {...})
```

File: src/components/picker/types.tsx [hunk 12]
```diff
@@ -147,49 +130,112 @@ export type PickerBaseProps = Omit<NewTextFieldProps, 'value' | 'onChange'> & {
...
    */
-  useWheelPicker?: boolean;
+  renderPickerOverlay?: (modalProps: RenderCustomModalProps) => React.ReactElement;
   /**
-   * Pass props to the list component that wraps the picker options (allows to control FlatList behavior)
+   * Adds blur effect to picker modal (iOS only)
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:145]:
```suggestion
   * Add blur effect to picker modal (iOS only)
```

File: src/components/picker/types.tsx [hunk 13]
```diff
@@ -147,49 +130,112 @@ export type PickerBaseProps = Omit<NewTextFieldProps, 'value' | 'onChange'> & {
...
-  pickerModalProps?: object;
+  renderCustomOverlayHeader?: (callbacks: {onDone?: () => void; onCancel?: () => void}) => React.ReactElement;
   /**
-   * Custom container style
-   */
-  containerStyle?: StyleProp<ViewStyle>;
-  /**
-   * @deprecated
-   * Callback for modal onShow event
+   * Pass props to the list component that wraps the picker options (allows to control FlatList behavior)
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:156]:
```suggestion
   * Pass props to the list component that wraps the picker items (allows to control FlatList behavior)
```

File: src/components/picker/types.tsx [hunk 14]
```diff
@@ -147,49 +130,112 @@ export type PickerBaseProps = Omit<NewTextFieldProps, 'value' | 'onChange'> & {
...
+    /**
+     * Picker current value in the shape of {value: ..., label: ...}, for custom shape use 'getItemValue' prop
+     */
+    value?: PickerValue;
+    /**
+     * Callback for when picker value change
+     */
+    onChange?: (value: PickerValue) => void;
+    /**
+     * SINGLE mode or MULTI mode
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:188]:
```suggestion
     * SINGLE or MULTI selection mode
```

File: src/components/picker/types.tsx [hunk 15]
```diff
@@ -147,49 +130,112 @@ export type PickerBaseProps = Omit<NewTextFieldProps, 'value' | 'onChange'> & {
...
+     */
+    getLabel?: (value: PickerValue) => string;
+    /**
+     * The picker modal top bar props
+     */
+    topBarProps?: ModalTopBarProps;
+    /**
+     * Use wheel picker instead of a list picker
+     */
+    useWheelPicker?: boolean;
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:224]: Please move up next to useDialog

File: src/components/picker/types.tsx (hunk 1 above)
Comment (adids1221) [reply src/components/picker/types.tsx:4]: reverted

File: src/components/picker/types.tsx [hunk 16]
```diff
@@ -147,49 +130,112 @@ export type PickerBaseProps = Omit<NewTextFieldProps, 'value' | 'onChange'> & {
...
+    /**
+     * Custom picker input container style
+     */
+    inputContainerStyle?: StyleProp<ViewStyle>;
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:228]: containerStyle is better - I wouldn't change the name here

File: src/components/picker/types.tsx (hunk 2 above)
Comment (adids1221) [reply src/components/picker/types.tsx:154]: renamed

File: src/components/picker/types.tsx (hunk 3 above)
Comment (adids1221) [reply src/components/picker/types.tsx:143]: renamed

File: src/components/picker/types.tsx [hunk 17]
```diff
@@ -147,49 +130,112 @@ export type PickerBaseProps = Omit<NewTextFieldProps, 'value' | 'onChange'> & {
...
+  PickerListProps & {
+    /* ...TextField.propTypes, */
+    /**
+     * Use dialog instead of modal picker
+     */
+    useDialog?: boolean;
+    /**
+     * Pass for different field type UI (form, filter or settings)
+     */
+    fieldType?: PickerFieldTypes | `${PickerFieldTypes}`;
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:178]: Please move before 'renderInput'

File: src/components/picker/types.tsx (hunk 4 above)
Comment (adids1221) [reply src/components/picker/types.tsx:220]: I'll move it to the PickerListProps.

File: src/components/picker/types.tsx (hunk 7 above)
Comment (adids1221) [reply src/components/picker/types.tsx:94]: I'll do this change in the new renamed prop - renderOverlay

File: src/components/picker/index.tsx [hunk 18]
```diff
@@ -264,22 +250,20 @@ const Picker = React.forwardRef((props: PickerProps, ref) => {
...
           {...customPickerProps}
           disabled={themeProps.editable === false}
         >
-          {renderPicker ? (
+          {renderInput ? (
             // @ts-expect-error - hopefully will be solved after the picker migration ends
-            renderPicker(value, label)
+            renderInput(value, label)
           ) : (
             <TextField
```
Comment (Inbal-Tish) [inline src/components/picker/index.tsx:268]: Let's move this part to a 'renderTextField' method to have it cleaner

File: src/components/picker/types.tsx (hunk 16 above)
Comment (adids1221) [reply src/components/picker/types.tsx:228]: renamed back to containerStyle

File: src/components/picker/types.tsx [hunk 19]
```diff
@@ -1,7 +1,7 @@
 import React, {PropsWithChildren, ReactNode} from 'react';
 import {FlatListProps, StyleProp, ViewStyle, TextStyle} from 'react-native';
 import {ExpandableOverlayProps, ExpandableOverlayMethods} from '../../incubator/expandableOverlay';
-import {ModalTopBarProps} from '../modal/TopBar';
+import {ModalTopBarProps} from '../../components/modal';
 // TODO: Replace with new TextField Props after migration to new TextField has completed
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:5]: Can we remove this TODO and the rename to 'NewTextFieldProps'?

File: src/components/picker/index.tsx [hunk 20]
```diff
@@ -208,10 +194,44 @@ const Picker = React.forwardRef((props: PickerProps, ref) => {
...
 
+  const renderTextField = useCallback(() => {
+    return renderInput ? (
+      // @ts-expect-error - hopefully will be solved after the picker migration ends
+      renderInput(value, label)
+    ) : (
+      <TextField
+        // @ts-expect-error
+        ref={pickerRef}
+        {...others}
```
Comment (Inbal-Tish) [inline src/components/picker/index.tsx:209]: others now includes all the props we are passing to the newProps hook. Let's make sure there is no prop that can affect the TextField here

File: src/components/picker/types.tsx [hunk 21]
```diff
@@ -48,85 +46,62 @@ export interface PickerSearchStyle {
...
+   * @deprecated
+   * A function that extract the label out of the value prop in case value has a custom structure (e.g. {myValue, myLabel})
    */
-  onChange?: (value: PickerValue) => void;
+  getItemLabel?: (value: PickerValue) => string;
   /**
-   * SINGLE mode or MULTI mode
+   * @deprecated
+   * Callback for modal onShow event
+   * Instead pass onShow via customPickerProps.modalProp | pickerModalProps
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:68]:
```suggestion
   * Instead pass onShow via customPickerProps.modalProps.onShow
```

File: src/components/picker/types.tsx [hunk 22]
```diff
@@ -48,85 +46,62 @@ export interface PickerSearchStyle {
...
+   * @deprecated
+   * instead pass items
    */
-  selectionLimit?: number;
+  children?: ReactNode | undefined;
   /**
-   * Adds blur effect to picker modal (iOS only)
+   * @deprecated
+   * Render a custom header for Picker's dialog
+   * instead use renderCustomOverlayHeader
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:79]:
```suggestion
   * instead use renderHeader
```

File: src/components/picker/types.tsx [hunk 23]
```diff
@@ -48,85 +46,62 @@ export interface PickerSearchStyle {
...
-  /**
-   * The picker modal top bar props
+   * Pass props to the picker modal
+   * instead pass modalProps via customPickerProps.modalProps
    */
-  topBarProps?: ModalTopBarProps;
+  pickerModalProps?: ExpandableOverlayProps['modalProps'];
+};
+
+type PickerSearchProps = {
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:104]: Please move before PickerListProps

File: src/components/picker/types.tsx [hunk 24]
```diff
@@ -48,85 +46,62 @@ export interface PickerSearchStyle {
...
-  customPickerProps?: ExpandableOverlayProps;
-  /**
-   * Add onPress callback for when pressing the picker
-   */
-  onPress?: () => void;
   /**
    * @deprecated
-   * A function that extract the unique value out of the value prop in case value has a custom structure (e.g. {myValue, myLabel})
+   * Render custom picker overlay (e.g ({visible, children, toggleModal}) => {...})
+   * instead use renderPickerOverlay
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:93]:
```suggestion
   * instead use renderOverlay
```

File: src/components/picker/types.tsx [hunk 25]
```diff
@@ -147,49 +122,111 @@ export type PickerBaseProps = Omit<NewTextFieldProps, 'value' | 'onChange'> & {
...
+  renderOverlay?: (modalProps: RenderCustomModalProps) => React.ReactElement;
   /**
-   * Pass props to the list component that wraps the picker options (allows to control FlatList behavior)
+   * Add blur effect to picker modal (iOS only)
    */
-  listProps?: Partial<FlatListProps<any>>;
+  enableModalBlur?: boolean;
+};
+
+type PickerListProps = PickerSearchProps & {
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:142]: What about 'PickerItemsListProps'? Is not a part of this type?

Comment (adids1221) [reply src/components/picker/types.tsx:142]: As we spoke I'll rearrange it but we will leave it this way.

File: src/components/picker/types.tsx [hunk 26]
```diff
@@ -147,49 +122,111 @@ export type PickerBaseProps = Omit<NewTextFieldProps, 'value' | 'onChange'> & {
...
+    /**
+     * Use dialog instead of modal picker
+     */
+    useDialog?: boolean;
+    /**
+     * Use wheel picker instead of a list picker
+     */
+    useWheelPicker?: boolean;
+    /**
+     * Picker current value in the shape of {value: ..., label: ...}, for custom shape use 'getItemValue' prop
```
Comment (Inbal-Tish) [inline src/components/picker/types.tsx:175]: getItemValue is deprecated...

Comment (adids1221) [reply src/components/picker/types.tsx:175]: Also the shape isn't right.
The type is string | number or array of that.
```suggestion
     * Picker value
```

File: src/components/picker/index.tsx (hunk 20 above)
Comment (adids1221) [reply src/components/picker/index.tsx:209]: Verified, it's ok

**Review by Inbal-Tish — APPROVED** [review]

**Comment by Inbal-Tish** [comment]
@adids1221 Approved. Let's run the tests here and then merge if we didn't break anything

**Comment by adids1221** [comment]
depends on [testing PR](https://github.com/wix/react-native-ui-lib/pull/3107)
